client/src/main/java/io/pravega/client/stream/impl/Pinger.java

```diff
@@ -89,14 +99,21 @@ private void startPeriodicPingTxn() {
      */
     private void pingTransactions() {
         log.info("Start sending transaction pings.");
-        txnList.stream().forEach(uuid -> {
+
+        List<UUID> stopPingTxns = new ArrayList<>();
```

Comment:
1. This is using `ArrayList` and `LinkedBlockingQueue`, both of which have O(n) runtime for all operations except `add`. I don't see you using any feature that requires this to be a List or a Queue (no ordering anywhere, no getFirst/Last), therefore I suggest you make both of these `HashSet`s and use set-specific methods on that class to do your computation.
2. I also suggest you use a common lock object (could be one of the sets) to synchronize access to these sets. This method executes at a fixed rate (not interval) which means it is possible that two or more executions overlap thus corrupting your state. 
3. Consider using `SequentialProcessor` from `common.concurrent` to avoid overlapping execution of requests if what I said above (at 2.) can cause issues.

Reply:
Done, changed to use `HashSet`. 
Also, `java.util.concurrent.ScheduledExecutorService#scheduleAtFixedRate` ensure that there are no concurrent executions of the command, pingTransactions(), hence will not use `SequentialProcessor`.